### PR TITLE
Add Kafka Producer Metrics and Kafka Buffer Metrics

### PR DESCRIPTION
**Pull request by dinujoh**
### Description
Add Kafka Producer Metrics and Kafka Buffer Metrics
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java [hunk 1]
```diff
@@ -69,28 +72,39 @@ public class KafkaCustomProducer<T> {
...
+    KafkaTopicProducerMetrics getTopicMetrics() {
+        return topicMetrics;
     }
 
     public void produceRawData(final byte[] bytes, final String key) {
         try {
             send(topicName, key, bytes).get();
+            topicMetrics.update(producer);
         } catch (Exception e) {
+            // TODO: Metrics
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java:107]: These are still TODO?

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        Map<MetricName, ? extends Metric> metrics = producer.metrics();
+        for (Map.Entry<MetricName, ? extends Metric> entry : metrics.entrySet()) {
+            MetricName metric = entry.getKey();
+            Metric value = entry.getValue();
+            String metricName = metric.name();
+            //if(metric.tags().containsKey("topic") && metricName.contains("lag")) {
+            if ((metricName.contains("send") || metricName.contains("byte")) && metric.tags().containsKey("topic")) {
+                System.out.println("Metric Name: " + metricName);
+                System.out.println("Metric Value: " + value.metricValue());
+                System.out.println("Metric Tag: " + metric.tags());
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java:102]: Should remove these print statements

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    public void update(final KafkaProducer producer) {
+        Map<String, Double> producerMetrics = metricValues.get(producer);
+
+        Map<MetricName, ? extends Metric> metrics = producer.metrics();
+        for (Map.Entry<MetricName, ? extends Metric> entry : metrics.entrySet()) {
+            MetricName metric = entry.getKey();
+            Metric value = entry.getValue();
+            String metricName = metric.name();
+            //if(metric.tags().containsKey("topic") && metricName.contains("lag")) {
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java:98]: Should remove this comment

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                   continue;
+                }
+
+                double newValue = (Double)value.metricValue();
+                if (metricName.equals("record-send-total")) {
+                    synchronized(producerMetrics) {
+                        double prevValue = producerMetrics.get(metricName);
+                        numberOfRecordsSent.increment(newValue - prevValue);
+                    }
+                } else if (metricName.equals("byte-total")) {
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java:116]: Lot of these keyword strings are used multiple times. Would be better to put them into variables.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java (hunk 1 above)
Comment (dinujoh) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java:107]: These are producer metrics for exception handling. Needs to be done as separate PR.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java (hunk 2 above)
Comment (dinujoh) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java]: Will fix.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java (hunk 3 above)
Comment (dinujoh) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java]: Will fix.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java (hunk 4 above)
Comment (dinujoh) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/util/KafkaTopicProducerMetrics.java:110]: Will fix.

**Review by graytaylor0 — DISMISSED** [review]

**Review by hshardeesi — APPROVED** [review]

**Review by chenqi0805 — APPROVED** [review]

**Review by graytaylor0 — APPROVED** [review]

**Pull request merged**